ddtrace/tracer/textmap.go

```diff
@@ -265,23 +266,60 @@ func (p *chainedPropagator) Inject(spanCtx ddtrace.SpanContext, carrier interfac
 	return nil
 }
 
-// Extract implements Propagator.
+// Extract implements Propagator. This method will attempt to extract the context
+// based on the precedence order of the propagators. Generally, the first valid
+// trace context that could be extracted will be returned, and other extractors will
+// be ignored. However, the W3C tracestate header value will always be extracted and
+// stored in the local trace context even if a previous propagator has already succeeded
+// so long as the trace-ids match.
 func (p *chainedPropagator) Extract(carrier interface{}) (ddtrace.SpanContext, error) {
+	var ctx ddtrace.SpanContext
 	for _, v := range p.extractors {
-		ctx, err := v.Extract(carrier)
-		if ctx != nil {
-			// first extractor returns
-			log.Debug("Extracted span context: %#v", ctx)
-			return ctx, nil
+		if p, isW3C := v.(*propagatorW3c); isW3C && ctx != nil {
+			// A valid trace context has already been extracted. We should now extract
+			// and propagate the W3C tracestate header if the trace-id matches.
+			p.propagateTracestate(ctx.(*spanContext), carrier)
+			sctx := ctx.(*spanContext)
+			fmt.Println(sctx)
+			break
 		}
-		if err == ErrSpanContextNotFound {
-			continue
+		var err error
+		ctx, err = v.Extract(carrier)
+		if ctx != nil {
+			if internal.BoolEnv("DD_TRACE_PROPAGATION_EXTRACT_FIRST", false) {
```

Comment:
This is a system call. Is there a way to avoid doing this every time?

Reply:
Yep 👍

Reply:
I see you added the `onlyExtractFirst` in the `chainedPropagator`, did you mean to also use it here?

Reply:
I did 😄